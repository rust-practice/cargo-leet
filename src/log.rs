use anyhow::Context;
use log::LevelFilter;
use log4rs::{
    append::console::ConsoleAppender,
    config::{Appender, Root},
    encode::pattern::PatternEncoder,
};

pub fn init_logging(log_level: LevelFilter) -> anyhow::Result<()> {
    let stdout = ConsoleAppender::builder()
        .encoder(Box::new(PatternEncoder::new("{h({M} {l} - {m})}{n}")))
        .build();

    let config = log4rs::Config::builder()
        .appender(Appender::builder().build("stdout", Box::new(stdout)))
        .build(Root::builder().appender("stdout").build(log_level))?;
    log4rs::init_config(config).context("Failed to initialize logging")?;
    Ok(())
}
