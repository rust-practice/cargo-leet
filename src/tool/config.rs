pub(crate) struct Config {}

impl Config {
    // assumed in the code using them URLs Must end with trailing "/"
    pub(crate) const LEETCODE_PROBLEM_URL: &str = "https://leetcode.com/problems/";
    pub(crate) const LEETCODE_GRAPH_QL: &str = "https://leetcode.com/graphql/";
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn problem_url_ends_with_slash() {
        assert!(Config::LEETCODE_PROBLEM_URL.ends_with('/'));
    }

    #[test]
    fn graph_ql_url_ends_with_slash() {
        assert!(Config::LEETCODE_GRAPH_QL.ends_with('/'));
    }
}
