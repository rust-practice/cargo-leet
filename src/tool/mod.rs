pub(crate) mod cli;
pub(crate) mod config;
pub(crate) mod core;
pub(crate) mod log;
