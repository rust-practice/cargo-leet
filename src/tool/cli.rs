use std::env;

use anyhow::Context;
use clap::{Args, Parser, Subcommand, ValueEnum};
use log::{info, LevelFilter};

/// Top level entry point for command line arguments parsing
///
/// Based on example <https://docs.rs/clap/latest/clap/_derive/_cookbook/cargo_example_derive/>
#[derive(Parser)]
#[command(name = "cargo")]
#[command(bin_name = "cargo")]
pub enum CargoCli {
    /// This is necessary because it a cargo subcommand so the first argument
    /// needs to be the command name
    Leet(Cli),
}
#[derive(Args, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    #[command(subcommand)]
    pub command: Commands,

    /// Specify the path to the project root (If not provided uses current
    /// working directory)
    #[arg(long, short, global = true, value_name = "FOLDER")]
    path: Option<String>,

    /// Set logging level to use
    #[arg(long, short, global = true, value_enum, default_value_t = LogLevel::Warn)]
    pub log_level: LogLevel,
}

impl Cli {
    /// Changes the current working directory to path if one is given
    pub fn update_current_working_dir(&self) -> anyhow::Result<()> {
        info!(
            "Before attempting update current dir, it is: {}",
            env::current_dir()?.display()
        );
        if let Some(path) = &self.path {
            info!("Going to update working directory to to '{path}'");
            std::env::set_current_dir(path)
                .with_context(|| format!("Failed to set current dir to: '{path}'"))?;
            info!(
                "After updating current dir, it is: '{}'",
                env::current_dir()?.display()
            );
        } else {
            info!("No user supplied path found. No change")
        }
        Ok(())
    }
}

#[derive(Subcommand, Debug)]
pub enum Commands {
    #[clap(visible_alias = "gen", short_flag = 'g')]
    Generate(GenerateArgs),
}

#[derive(Args, Debug)]
pub struct GenerateArgs {
    /// Question slug or url (If none specified then daily challenge is used)
    pub problem: Option<String>,
    /// If set the module name generated includes the number for the problem
    #[arg(short = 'n', long = "number_in_name", default_value_t = false)]
    pub should_include_problem_number: bool,
}

/// Exists to provide better help messages variants copied from LevelFilter as
/// that's the type that is actually needed
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, ValueEnum, Debug)]
pub enum LogLevel {
    /// Nothing emitted in this mode
    Off,
    Error,
    Warn,
    Info,
    Debug,
    Trace,
}

impl From<LogLevel> for LevelFilter {
    fn from(value: LogLevel) -> Self {
        match value {
            LogLevel::Off => LevelFilter::Off,
            LogLevel::Error => LevelFilter::Error,
            LogLevel::Warn => LevelFilter::Warn,
            LogLevel::Info => LevelFilter::Info,
            LogLevel::Debug => LevelFilter::Debug,
            LogLevel::Trace => LevelFilter::Trace,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn verify_cli() {
        // Source: https://docs.rs/clap/latest/clap/_derive/_tutorial/index.html#testing
        // My understanding it reports most development errors without additional effort
        use clap::CommandFactory;
        CargoCli::command().debug_assert()
    }
}
