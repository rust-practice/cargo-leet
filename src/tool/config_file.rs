use std::fs;

use anyhow::Context;
use log::info;
use serde::{Deserialize, Serialize};

#[derive(Clone, Serialize, Deserialize, Debug, Default)]
#[serde(default, deny_unknown_fields)]
pub(crate) struct ConfigFile {
    pub active: Option<String>,
    pub should_include_problem_number_in_mod_name: bool,
}

impl ConfigFile {
    const FILENAME: &str = ".leet.toml";

    pub(crate) fn load() -> anyhow::Result<Self> {
        let content = match std::fs::read_to_string(Self::FILENAME) {
            Ok(x) => x,
            Err(e) => {
                info!(
                    "failed to load {:?}. Using defaults. Error msg: {e:?}",
                    Self::FILENAME
                );
                return Ok(Self::default());
            }
        };

        toml::from_str(&content).with_context(|| format!("failed to parse {}", Self::FILENAME))
    }

    pub(crate) fn save(&self) -> anyhow::Result<()> {
        let content = toml::to_string(&self).context("failed to convert toml")?;
        fs::write(Self::FILENAME, content)
            .with_context(|| format!("failed to write {}", Self::FILENAME))?;
        Ok(())
    }
}
