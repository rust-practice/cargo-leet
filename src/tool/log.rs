#![cfg(feature = "tool")]
use env_logger::Builder;
use log::LevelFilter;

pub fn init_logging(level: LevelFilter) -> anyhow::Result<()> {
    Builder::new().filter(None, level).try_init()?;
    Ok(())
}
