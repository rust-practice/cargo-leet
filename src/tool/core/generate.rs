use anyhow::{bail, Context};
use convert_case::{Case, Casing};
use log::info;
use std::borrow::Cow;

use crate::tool::{
    cli,
    config::Config,
    core::helpers::{
        code_snippet::get_code_snippet_for_problem, daily_challenge,
        problem_metadata::get_problem_metadata, write_to_disk,
    },
};

pub(crate) fn do_generate(args: &cli::GenerateArgs) -> anyhow::Result<()> {
    let title_slug: Cow<String> = if let Some(specific_problem) = &args.problem {
        get_slug_from_args(specific_problem)
            .with_context(|| format!("Expected URL or slug but got {specific_problem}"))?
    } else {
        // Daily problem
        let slug = daily_challenge::get_daily_challenge_slug()?;
        info!("Slug for daily problem is: '{slug}'");
        Cow::Owned(slug)
    };

    let (module_name, module_code) = create_module_code(title_slug, args)
        .context("Failed to generate the name and module code")?;
    write_to_disk::write_file(&module_name, module_code).context("Failed to write to disk")?;
    println!("Generated module: {module_name}");
    Ok(())
}

fn get_slug_from_args(specific_problem: &String) -> anyhow::Result<Cow<'_, String>> {
    Ok(if is_url(specific_problem) {
        // Working with a url
        info!("Using '{specific_problem}' as a url");
        let slug = url_to_slug(specific_problem)?;
        info!("Extracted slug '{slug}' from url");
        Cow::Owned(slug)
    } else {
        // This is expected to be a valid slug
        //TODO: Validate slug
        info!("Using '{specific_problem}' as a slug");
        Cow::Borrowed(specific_problem)
    })
}

/// Gets the code and other data from leetcode and generates the suitable code for the module and the name of the module
/// Returns the module name and the module code
///
/// NB: Did not return `Cow` because `module_name` is always a modified version of the input
pub fn create_module_code(
    title_slug: Cow<String>,
    args: &cli::GenerateArgs,
) -> anyhow::Result<(String, String)> {
    info!("Building module contents for {title_slug}");

    let meta_data =
        get_problem_metadata(&title_slug).context("Failed to retrieve problem meta data")?;

    // Add problem URL
    let mut code_snippet = format!(
        "//! Solution for {}{title_slug}\n",
        Config::LEETCODE_PROBLEM_URL
    );

    // Add problem number and title
    code_snippet.push_str(&format!(
        "//! {}\n",
        meta_data
            .get_num_and_title()
            .context("Failed to get problem number and title")?
    ));

    // Add blank line between docstring and code
    code_snippet.push('\n');

    // Get code snippet
    let problem_code = get_code_snippet_for_problem(&title_slug)?;
    code_snippet.push_str(problem_code.as_ref());

    // Add 2 empty lines between code and "other stuff (like tests and struct definition"
    code_snippet.push_str("\n\n");

    // Add struct for non design questions
    if problem_code.type_.is_non_design() {
        code_snippet.push_str("\npub struct Solution;\n")
    }

    // Add leet code types
    if problem_code.has_tree() {
        code_snippet.push_str("use cargo_leet::TreeNode;\n");
    }
    if problem_code.has_list() {
        code_snippet.push_str("use cargo_leet::ListNode;\n");
    }

    // Add tests
    let tests = meta_data.get_test_cases(&problem_code)?;
    code_snippet.push_str(&tests);

    // Set module name
    let module_name = if args.should_include_problem_number {
        info!("Including problem number in module name");
        format!(
            "_{}_{}",
            meta_data.get_id()?,
            title_slug.to_case(Case::Snake)
        )
    } else {
        info!("Using snake case slug for module name");
        title_slug.to_case(Case::Snake)
    };

    Ok((module_name, code_snippet))
}

/// Quick and dirty test to see if this is a url
fn is_url(value: &str) -> bool {
    value.contains('/')
}

fn url_to_slug(url: &str) -> anyhow::Result<String> {
    assert!(Config::LEETCODE_PROBLEM_URL.ends_with('/'));
    if !url.starts_with(Config::LEETCODE_PROBLEM_URL) {
        bail!(
            "Expected a leetcode url that starts with '{}' but got '{url}'",
            Config::LEETCODE_PROBLEM_URL
        )
    }
    let split_url: Vec<_> = url.split('/').collect();
    let split_prefix: Vec<_> = Config::LEETCODE_PROBLEM_URL.split('/').collect();
    debug_assert!(split_prefix.len() < split_url.len());
    Ok(split_url[split_prefix.len() - 1].to_string())
}
