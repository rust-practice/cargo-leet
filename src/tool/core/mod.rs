mod generate;
mod helpers;

use self::generate::do_generate;
use crate::tool::cli::{self, Cli};
use anyhow::{bail, Context};
use std::{env, path::Path};

/// Entry point used by the tool. The `main.rs` is pretty thin shim around this
/// function.
///
/// # Errors
/// - Current directory does not exist.
/// - There are insufficient permissions to access the current directory.
/// - Returns an Err if the operation fails.
pub fn run(cli: &Cli) -> anyhow::Result<()> {
    cli.update_current_working_dir()?;

    working_directory_validation()?;

    match &cli.command {
        cli::Commands::Generate(args) => do_generate(args),
    }
}

fn working_directory_validation() -> anyhow::Result<()> {
    let req_file = "Cargo.toml";
    let path = Path::new(req_file);
    if path.exists() {
        Ok(())
    } else {
        bail!(
            "Failed to find {req_file} in current directory '{}'",
            env::current_dir()
                .context("Failed to get current working directory")?
                .display()
        );
    }
}
