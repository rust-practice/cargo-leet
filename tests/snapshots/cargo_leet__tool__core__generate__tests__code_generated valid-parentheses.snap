---
source: src/tool/core/generate.rs
expression: code_generated
---
//! Solution for https://leetcode.com/problems/valid-parentheses
//! 20. Valid Parentheses

impl Solution {
    pub fn is_valid(s: String) -> bool { todo!("Fill in body") }
}

// << ---------------- Code below here is only for local use ---------------- >>

pub struct Solution;

#[cfg(test)]
mod tests {
    use super::*;
    

    use rstest::rstest;

    #[rstest]
    #[case("()", true)]
    #[case("()[]{}", true)]
    #[case("(]", false)]
    #[case("([])", true)]
    fn case(#[case] s: String, #[case] expected: bool) {
        let actual = Solution::is_valid(s);
        assert_eq!(actual, expected);
    }
}
